### PR TITLE
auto-close if duration is a positive finite number

### LINKED ISSUE #23: Paper-toast should not hide on mouseover
When you have a paper-toast requesting an action such "This didn't work. Retry" or "Your store is live. View your store" the toast should not disappear as the user is hovering over the toast.

### LINKED ISSUE #27: Paper-toast permanent duration
It would be cool to be able to set an attribute for an infinite duration to use the paper-toast for things like showing a terms and cookie policy link on a site, shown [here](https://silktide.com/the-stupid-cookie-law-is-dead-at-last/) which can't go away until it is closed by the user.
You can already just put in a huge wait so it's necessary, It would just be cool.

Maybe something like

    <paper-toast permanent></paper-toast>

or even possible

    <paper-toast wait="0"></paper-toast>

I'm not sure what actually happens when you put a wait time of '0', except it calls debounce with a wait of '0' which calls async with a wait time of '0'

paper-toast.html
```javascript
    show: function() {
      if (PaperToast.currentToast) {
        PaperToast.currentToast.hide();
      }
      PaperToast.currentToast = this;
      this.removeAttribute('aria-hidden');
      this._setVisible(true);
      this.fire('iron-announce', {
        text: this.text
      });
      this.debounce('hide', this.hide, this.duration);
    },
````
polymer.html
```javascript
    Polymer.Async = {
        _currVal: 0,
        _lastVal: 0,
        _callbacks: [],
        _twiddleContent: 0,
        _twiddle: document.createTextNode(''),
        run: function(callback, waitTime) {

            //- if (waitTime === 0) possibly just return so the function never gets run
            if (waitTime > 0) {

                //- Not relevant to this issue, but what does the ~ do to the return value of setTimeout?
                return ~setTimeout(callback, waitTime);
            } else {
                this._twiddle.textContent = this._twiddleContent++;
                this._callbacks.push(callback);
                return this._currVal++;
            }
        },
        cancel: function(handle) {
            if (handle < 0) {
                clearTimeout(~handle);
            } else {
                var idx = handle - this._lastVal;
                if (idx >= 0) {
                    if (!this._callbacks[idx]) {
                        throw 'invalid async handle: ' + handle;
                    }
                    this._callbacks[idx] = null;
                }
            }
        },
        _atEndOfMicrotask: function() {
            var len = this._callbacks.length;
            for (var i = 0; i < len; i++) {
                var cb = this._callbacks[i];
                if (cb) {
                    try {
                        cb();
                    } catch (e) {
                        i++;
                        this._callbacks.splice(0, i);
                        this._lastVal += i;
                        this._twiddle.textContent = this._twiddleContent++;
                        throw e;
                    }
                }
            }
            this._callbacks.splice(0, len);
            this._lastVal += len;
        }
    };
    new(window.MutationObserver || JsMutationObserver)(Polymer.Async._atEndOfMicrotask.bind(Polymer.Async)).observe(Polymer.Async._twiddle, {
        characterData: true
    });
    Polymer.Debounce = function() {
        var Async = Polymer.Async;
        var Debouncer = function(context) {
            this.context = context;
            this.boundComplete = this.complete.bind(this);
        };
        Debouncer.prototype = {
            go: function(callback, wait) {
                var h;
                this.finish = function() {
                    Async.cancel(h);
                };
                h = Async.run(this.boundComplete, wait);
                this.callback = callback;
            },
            stop: function() {
                if (this.finish) {
                    this.finish();
                    this.finish = null;
                }
            },
            complete: function() {
                if (this.finish) {
                    this.stop();
                    this.callback.call(this.context);
                }
            }
        };

        function debounce(debouncer, callback, wait) {
            if (debouncer) {
                debouncer.stop();
            } else {
                debouncer = new Debouncer(this);
            }
            debouncer.go(callback, wait);
            return debouncer;
        }
        return debounce;
    }();
```

:cat:

### PR DESCRIPTION
**Pull request by valdrinkoshi**
Fixes #23, fixes #27 by calling `hide` only if `duration` is a positive finite number.
If `duration` is `<= 0` or `Infinite`, the toast will stay opened until `hide()` is called, or another instance of `paper-toast` is opened

### CODE REVIEW

**Comment by valdrinkoshi** [comment]
@cdata FYI this PR is done on top of the commit for PR #34 (that's why we now see 2 commits). After #34 is merged, we'll see 1 commit only in here.

**Comment by valdrinkoshi** [comment]
@cdata kindly review.
This PR goes after #34.
The diffs between the 2 branches here https://github.com/PolymerElements/paper-toast/compare/issue-13...issue-27

File: bower.json [hunk 1]
```diff
@@ -24,10 +24,11 @@
     "iron-a11y-announcer": "polymerelements/iron-a11y-announcer#^1.0.0"
   },
   "devDependencies": {
-    "iron-component-page": "polymerelements/iron-component-page#^1.0.0",
-    "paper-button": "polymerelements/paper-button#^1.0.0",
     "web-component-tester": "*",
```
Comment (cdata) [inline bower.json]: This is orthogonal to this change, but could you do me a favor and update this to: `"web-component-tester": "polymer/web-component-tester#^3.4.0"`

:bow:

File: paper-toast.html [hunk 2]
```diff
@@ -112,10 +111,13 @@
       Polymer.IronA11yAnnouncer.requestAvailability();
     },
 
-    ready: function() {
-      this.async(function() {
+    attached: function() {
+      var visible = (this.hasAttribute('visible') && this.getAttribute('visible') !== 'false');
```
Comment (cdata) [inline paper-toast.html]: Can we just measure `this.visible`?

Comment (valdrinkoshi) [inline paper-toast.html]: `this.visible` is read-only, hence it won't be set, so we have to check for the attribute here

File: bower.json (hunk 1 above)
Comment (valdrinkoshi) [inline bower.json]: Sure, I'll do it in PR #34 though (this PR is based on that commit)

**Comment by valdrinkoshi** [comment]
Alright, PR ready for review @cdata :)

File: test/basic.html [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+    <test-fixture id="show">
+      <template>
+        <paper-toast></paper-toast>
+      </template>
+    </test-fixture>
+
+    <script>
+
+      suite('basic', function() {
```
Comment (cdata) [inline test/basic.html]: :100: Thanks so much for adding a test suite!

File: bower.json [hunk 4]
```diff
@@ -24,10 +24,10 @@
     "iron-a11y-announcer": "polymerelements/iron-a11y-announcer#^1.0.0"
   },
   "devDependencies": {
+    "web-component-tester": "polymer/web-component-tester#^3.4.0",
```
Comment (cdata) [inline bower.json]: :+1:

File: paper-toast.html [hunk 5]
```diff
@@ -132,7 +129,10 @@
       this.fire('iron-announce', {
         text: this.text
       });
-      this.debounce('hide', this.hide, this.duration);
+      // auto-close if duration is a positive finite number
+      if (this.duration > 0 && this.duration !== Infinity) {
```
Comment (cdata) [inline paper-toast.html]: :sparkler:

**Comment by cdata** [comment]
LGTM :+1:
